### PR TITLE
Feature/validate token

### PR DESCRIPTION
**Pull request by turegjorup**
* Added `Firebase:JWT` to validate token
* Added helper methods to validate token
* Enforce use of `state` and `nonce`
* Added `Mockery` to enable mocking of static methods in `JWT`
* Updated Readme

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -38,15 +54,50 @@ use ItkDev\OpenIdConnect\Security\OpenIdConfigurationProvider;
 
 $provider = new OpenIdConfigurationProvider([
     'redirectUri' => 'https://some.url', // Absolute url to where the user is redirected after a successful login            
-    'urlConfiguration' => 'https:/.../openid-configuration', // url to OpenId Discovery document
-    'cachePath' => '/some/directory/openId-cache.php', // Path for caching above discovery document
+    'openIDConnectMetadataUrl' => 'https:/.../openid-configuration', // url to OpenId Discovery document
+    'cacheItemPool' => 'Psr6/CacheItemPoolInterface', // Pool for caching above discovery document
```
Comment (lats1) [inline README.md]: Small thing, but maybe a comment about this should be an instance of a class implementing the interface.

File: src/Security/OpenIdConfigurationProvider.php [hunk 2]
```diff
@@ -72,81 +229,148 @@ public function getResourceOwnerDetailsUrl(AccessToken $token): string
...
-                $this->refreshCache();
+        $cacheKey = self::CACHE_KEY . 'configuration';
+        try {
+            $item = $this->cacheItemPool->getItem($cacheKey);
+            if ($item->isHit()) {
+                $configuration = $item->get();
+            } else {
+                $configuration = $this->fetchJsonResource($this->openIDConnectMetadataUrl);
+                $item->set($configuration);
+                $item->expiresAfter($this->cacheDuration);
```
Comment (lats1) [inline src/Security/OpenIdConfigurationProvider.php:376]: Do you need to use the save-method on the CacheItemPool in order to persist the item?

**Review by lats1 — CHANGES_REQUESTED** [review]

File: README.md [hunk 3]
```diff
@@ -38,15 +54,50 @@ use ItkDev\OpenIdConnect\Security\OpenIdConfigurationProvider;
...
+unique string that can be used to identify the origin of the request."_
+
+See: [Send authentication requests](https://docs.microsoft.com/en-us/azure/active-directory-b2c/openid-connect#send-authentication-requests)
+
+You must persist these locally so that they can be used to validate the token
+when the user is redirected back to your application.
+
+```php
+// Get "state" and "nonce"
+$state = $provider->getState();
```
Comment (jekuaitk) [inline README.md]: Isn't  "state" only given a value once we have called 'getAuthorizationUrl'?

Comment (turegjorup) [reply README.md]: Good catch. Changed to use `getRandomState()`

File: src/Security/OpenIdConfigurationProvider.php (hunk 2 above)
Comment (turegjorup) [reply src/Security/OpenIdConfigurationProvider.php:376]: Documentation is unclear, but have found other examples calling `save()`. Added `$this->cacheItemPool->save($item);`

File: README.md (hunk 1 above)
Comment (turegjorup) [reply README.md]: Attempted clarification

File: README.md (hunk 3 above)
Comment (jekuaitk) [reply README.md]: `getRandomState()` is protected in AbstractProvider, possibly add `generateState()` function similar to  `generateNonce()`?

Comment (turegjorup) [reply README.md]: Thank you. Done.

**Review by jekuaitk — APPROVED** [review]

**Pull request merged**